editor/utils/dom.js

```diff
@@ -77,17 +145,101 @@ export function placeCaretAtEdge( container, start = false ) {
 		return;
 	}
 
+	if ( ! container.isContentEditable ) {
+		container.focus();
+		return;
+	}
+
 	// Content editables
 	const range = document.createRange();
 	range.selectNodeContents( container );
-	range.collapse( start );
+	range.collapse( reverse );
 	const sel = window.getSelection();
 	sel.removeAllRanges();
 	sel.addRange( range );
 	container.focus();
 }
 
 /**
+ * Polyfill.
+ * Get a collapsed range for a given point.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Document/caretRangeFromPoint
+ *
+ * @param  {Document} doc The document of the range.
+ * @param  {Float}    x   Horizontal position within the current viewport.
+ * @param  {Float}    y   Vertical position within the current viewport.
+ * @return {?Range}       The best range for the given point.
+ */
+function caretRangeFromPoint( doc, x, y ) {
+	if ( doc.caretRangeFromPoint ) {
+		return doc.caretRangeFromPoint( x, y );
+	}
+
+	if ( ! doc.caretPositionFromPoint ) {
+		return null;
+	}
+
+	const point = doc.caretPositionFromPoint( x, y );
+	const range = doc.createRange();
+
+	range.setStart( point.offsetNode, point.offset );
+	range.collapse( true );
+
+	return range;
+}
+
+/**
+ * Places the caret at the top or bottom of a given element.
+ *
+ * @param {Element} container Focusable element.
+ * @param {Boolean} reverse   True for bottom, false for top.
+ * @param {Boolean} noScroll  Set to true to prevent scrolling.
+ */
+export function placeCaretAtVerticalEdge( container, reverse, noScroll ) {
+	const rect = firstVerticalRect;
+
+	if ( ! rect || ! container.isContentEditable ) {
+		placeCaretAtHorizontalEdge( container, reverse );
+		return;
+	}
+
+	const buffer = rect.height / 2;
+	const editableRect = container.getBoundingClientRect();
+	const x = rect.left + ( rect.width / 2 );
+	const y = reverse ? ( editableRect.bottom - buffer ) : ( editableRect.top + buffer );
+	const selection = window.getSelection();
+
+	// Temporary high z-index above toolbars.
+	container.style.zIndex = '10000';
+
+	const range = caretRangeFromPoint( document, x, y );
+
+	container.style.zIndex = null;
+
+	if ( ! range || ! container.contains( range.startContainer ) ) {
+		if ( ! noScroll ) {
+			// Might be out of view.
+			// Easier than attempting to calculate manually.
+			container.scrollIntoView( reverse );
+			placeCaretAtVerticalEdge( container, reverse, true );
+			return;
+		}
+
+		placeCaretAtHorizontalEdge( container, reverse );
+		return;
+	}
+
+	selection.removeAllRanges();
+	selection.addRange( range );
+	container.focus();
+	// Editable was already focussed, it goes back to old range...
+	// This fixes it.
+	selection.removeAllRanges();
+	selection.addRange( range );
```

Comment:
Could you provide more detail on the need to repeat this? Is it a quirk?, is it something otherwise preventable?, etc.

Reply:
I suspect it might be related to TinyMCE's `editor.selection.lastFocusBookmark`, but I need to investigate. Maybe @ephox-mogran or @EphoxJames know more.

Reply:
I don't know off the top of my head but I will dig into the code and see if I can find out.

Reply:
Thanks! Ideally we wouldn't have to touch the TinyMCE editor instance at all here.